src/main/java/net/minecraftforge/common/data/BrewingRecipeBuilder.java

```diff
@@ -0,0 +1,407 @@
+    {
+        this.advancement.addCriterion(key, criterion);
+        return (T) this;
+    }
+
+    /**
+     * Save the recipe with the given key
+     *
+     * @param recipe the consumer to which the built recipe will be passed
+     * @param key    the key to save the recipe under
+     */
+    public void save(Consumer<FinishedRecipe> recipe, String key)
+    {
+        this.save(recipe, new ResourceLocation(key));
+    }
+
+    /**
+     * Save the recipe with the given location/id
+     *
+     * @param recipe   the consumer to which the built recipe will be passed
+     * @param location the location to save the recipe under
+     */
+    public void save(Consumer<FinishedRecipe> recipe, ResourceLocation location)
+    {
+        this.ensureValid(location);
+        this.advancement.parent(new ResourceLocation("recipes/root")).addCriterion("has_the_recipe", RecipeUnlockedTrigger.unlocked(location)).rewards(AdvancementRewards.Builder.recipe(location)).requirements(RequirementsStrategy.OR);
+        recipe.accept(build(location));
+    }
+
+    /**
+     * Build the recipe with the given location
+     *
+     * @param location the location/id of the recipe
+     */
+    protected abstract FinishedRecipe build(ResourceLocation location);
+
+    private void ensureValid(ResourceLocation location)
+    {
+        if (this.advancement.getCriteria().isEmpty())
+        {
+            throw new IllegalStateException("No way of obtaining recipe " + location);
+        }
+    }
+
+    public static class NormalBrewingRecipeBuilder extends BrewingRecipeBuilder
+    {
+        private final Ingredient base;
+        private final Ingredient reagent;
+        private final ItemStack result;
+
+        public NormalBrewingRecipeBuilder(Ingredient base, Ingredient reagent, ItemStack result)
+        {
+            super(ForgeMod.BREWING_SERIALIZER.get());
+            this.base = base;
+            this.reagent = reagent;
+            this.result = result;
+        }
+
+        @Override
+        protected FinishedRecipe build(final ResourceLocation location)
+        {
+            return new Result(location, this.type, this.base, this.reagent, this.result, this.advancement, new ResourceLocation(location.getNamespace(), "recipes/brewing/" + location.getPath()));
+        }
+
+        public record Result(ResourceLocation id,
+                             RecipeSerializer<?> type,
+                             Ingredient base,
+                             Ingredient reagent,
+                             ItemStack result,
+                             Advancement.Builder advancement,
+                             ResourceLocation advancementId) implements FinishedRecipe
+        {
+
+            @Override
+            public void serializeRecipeData(JsonObject tag)
+            {
+                final ResourceLocation result = ForgeRegistries.ITEMS.getKey(this.result.getItem());
+                if (result == null)
```

Comment:
Validation should not be happening this late.  It should be given before the `IFinishedRecipe` is constructed.